starknet_py/net/models/transaction.py

```diff
@@ -1,29 +1,224 @@
+    compress_program_post_dump,
+    decompress_program,
+)
+
+from starknet_py.cairo.selector import get_selector_from_name
+from starknet_py.net.client_models import TransactionType
+from starknet_py.net.models import compute_address
+from starknet_py.net.models.chains import StarknetChainId
+from starknet_py.net.schemas.common import Felt, NoneFelt
+from starknet_py.utils.crypto.transaction_hash import (
     TransactionHashPrefix,
-    calculate_transaction_hash_common,
+    compute_declare_transaction_hash,
+    compute_deploy_account_transaction_hash,
+    compute_transaction_hash,
 )
 
-# noinspection PyPep8Naming
-from starkware.starknet.definitions.transaction_type import TransactionType as TT
-from starkware.starknet.public.abi import get_selector_from_name
 
-# noinspection PyPep8Naming
-from starkware.starknet.services.api.gateway.transaction import Declare as DCL
-from starkware.starknet.services.api.gateway.transaction import DeployAccount as DAC
-from starkware.starknet.services.api.gateway.transaction import InvokeFunction as IF
-from starkware.starknet.services.api.gateway.transaction import Transaction as T
+@dataclass(frozen=True)
+class Transaction(ABC):
+    """
+    StarkNet transaction base class.
+    """
 
-from starknet_py.net.models.chains import StarknetChainId
-from starknet_py.utils.crypto.facade import pedersen_hash
-from starknet_py.utils.docs import as_our_module
+    version: int = field(metadata={"marshmallow_field": Felt()})
+
+    @property
+    @abstractmethod
+    def type(self) -> TransactionType:
+        """
+        Returns the corresponding TransactionType enum.
+        """
 
-Invoke = InvokeFunction = as_our_module(IF)
-Transaction = as_our_module(T)
-TransactionType = as_our_module(TT)
-Declare = as_our_module(DCL)
-DeployAccount = as_our_module(DAC)
+    @abstractmethod
+    def calculate_hash(self, chain_id: StarknetChainId) -> int:
+        """
+        Calculates the transaction hash in the StarkNet network - a unique identifier of the
+        transaction. See :ref:`compute_transaction_hash` docstring for more details.
+        """
+
+
+@dataclass(frozen=True)
+class AccountTransaction(Transaction):
+    """
+    Represents a transaction in the StarkNet network that is originated from an action of an
+    account.
+    """
+
+    max_fee: int = field(metadata={"marshmallow_field": Felt()})
+    signature: List[int] = field(
+        metadata={"marshmallow_field": fields.List(fields.String())}
+    )
+    nonce: int = field(metadata={"marshmallow_field": Felt()})
+
+
+# Used in the Account instead of Union[Invoke, Declare, DeployAccount]
+TypeAccountTransaction = TypeVar("TypeAccountTransaction", bound=AccountTransaction)
+
+
+@dataclass(frozen=True)
+class Declare(AccountTransaction):
+    """
+    Represents a transaction in the StarkNet network that is a declaration of a StarkNet contract
+    class.
+    """
+
+    contract_class: ContractClass
```

Comment:
One think to keep in mind: eventually we'll have to replace ContractClass with our own implementation.

Reply:
There is an existing issue for that.

Reply:
Right. #661